### ISSUE TITLE
`ReplaceOperation` is not applied correctly when value is an empty array

### ISSUE THREAD

**Issue by KMayne** [pingidentity/scim2#209]
**Describe the bug**
Applying a `ReplaceOperation` with an empty array value does not have any effect.

**To Reproduce**
```java
package scim2libbug;

import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.unboundid.scim2.common.exceptions.ScimException;
import com.unboundid.scim2.common.messages.PatchOperation;
import com.unboundid.scim2.common.types.Email;
import com.unboundid.scim2.common.types.UserResource;

public class Main {
    public static void main(String[] args) throws ScimException {
        var scimUser = new UserResource();
        var workEmail = new Email();
        workEmail.setPrimary(true);
        workEmail.setType("work");
        workEmail.setValue("test@email.com");
        scimUser.setEmails(workEmail);
        var genericUser = scimUser.asGenericScimResource();
        var patchOp = PatchOperation.replace("emails", JsonNodeFactory.instance.arrayNode());

        System.out.println("BEFORE:");
        System.out.println(genericUser);
        patchOp.apply(genericUser.getObjectNode());
        System.out.println("AFTER:");
        System.out.println(genericUser);
    }
}

/* Output:
BEFORE:
{
  "schemas" : [ "urn:ietf:params:scim:schemas:core:2.0:User" ],
  "emails" : [ {
    "value" : "test@email.com",
    "type" : "work",
    "primary" : true
  } ]
}
AFTER:
{
  "schemas" : [ "urn:ietf:params:scim:schemas:core:2.0:User" ],
  "emails" : [ {
    "value" : "test@email.com",
    "type" : "work",
    "primary" : true
  } ]
}
*/
```

**Expected behavior**
In the above example, the emails should be set to an empty list. I believe this was the behaviour that was intended to work in https://github.com/pingidentity/scim2/issues/103

**Additional context**
Looks like there's a condition in `JsonUtils` that's preventing the operation from applying:
https://github.com/pingidentity/scim2/blob/7dbeb1cacbebe48ccc5b221d3260153b3ae41160/scim2-sdk-common/src/main/java/com/unboundid/scim2/common/utils/JsonUtils.java#L319-L326

We're currently working around this for now by clearing array fields before applying our patch operations.

**Comment by kqarryzada** [comment]
@KMayne, thanks for your patience on this, as well as for the detail. I have a preliminary set of changes to address this issue, and plan to commit this in the next few weeks.

**Comment by KMayne** [comment]
> @KMayne, thanks for your patience on this, as well as for the detail. I have a preliminary set of changes to address this issue, and plan to commit this in the next few weeks.

Hi, I was discussing this with a colleague the other week and the spec is maybe ambiguous here.
The relevant section in the RFC is https://datatracker.ietf.org/doc/html/rfc7643#section-2.5

Personally I think my interpretation is correct as the spec goes on to say
> Assigning an attribute with the value "null" or an empty array (in the case of multi-valued attributes) has the effect of making the attribute "unassigned".

The confusion arises in that PATCH has slightly different semantics

There is discussion in the IETF mailing list about this since the spec was written
https://mailarchive.ietf.org/arch/msg/scim/jeLhV_MLcNfjkrld0Ki6D62yE1k/

**Comment by kqarryzada** [comment]
I agree with your interpretation that the highlighted sentence makes it clear that a replace operation with an empty array as its value indicates that the attribute on the existing resource should be removed. 

Phil Hunt's reply on that email chain, where he mentions that "[y]ou really have to test major implementers" to understand the de facto behavior also reaffirms this, in my opinion. In our Ping Identity Directory Server, we exhibit this behavior for SCIM 2 replace operations, and I'm pretty sure this is also commonly seen elsewhere (like in Azure).

**Comment by KMayne** [comment]
Nice one, thanks!

**Comment by kqarryzada** [comment]
@KMayne, this fix is available in the 3.1.0 release, which is available now.
